### DISCUSSION TITLE
Share your ideas!

### DISCUSSION THREAD

**Discussion by oskvr37** [oskvr37/tiddl#23] [Ideas]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪

**Comment by Darkloke** [comment]
Greetings. Can i ask?

1. This project is a replacement for tidal-dl by yaronzz, which seems to be abandoned for now?
2. Is it available on windows platform?

**Reply by oskvr37** [reply]
Hi, tidal-dl is likely abandoned by its author and tiddl is aspiring to be replacement for tidal-dl.
This project is available on every platform that can run Python3.

**Reply by Darkloke** [reply]
I see, thank you. Can i also ask the following?

While using tidal-dl i had two major problems (for me) with downloads.
1. Sometimes tidal-dl downloaded files with empty metadata and even if you re-download this files again - it just say its already exist, so nothing changes. I had to open such files in MP3tag, check for empty metadata, delete them and then re-download them again. Have you encountered similar behavior in your version?
2. Does it work via proxy?

**Reply by oskvr37** [reply]
You have to install tiddl and try yourself, I dont think that this problem with metadata occurs here.
I have never used tiddl with proxy and I can't answer this question, but if you used tidal-dl with proxy then it should work here too.

**Reply by Darkloke** [reply]
Can you tell me, please, how to run it on windows? I am ashamed to say i had no python experience.
1. I installed python 3.13.0
2. I installed the script with `pip install tiddl`
3. Now if i type 'tiddl' in cmd i get the following:
`Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "C:\Users\Darkloke\AppData\Local\Programs\Python\Python313\Scripts\tiddl.exe\__main__.py", line 4, in <module>
    from tiddl import main
  File "C:\Users\Darkloke\AppData\Local\Programs\Python\Python313\Lib\site-packages\tiddl\__init__.py", line 6, in <module>
    from .api import TidalApi
  File "C:\Users\Darkloke\AppData\Local\Programs\Python\Python313\Lib\site-packages\tiddl\api.py", line 2, in <module>
    from requests import Session
ModuleNotFoundError: No module named 'requests'`

**Reply by oskvr37** [reply]
Just a quick note: as the project owner, I’m here to answer questions specific to the project itself, rather than basic Python questions.
In this case, the error trace you shared points to a missing package, which you can install with:
```bash
pip install requests
```
For general Python questions, I would recommend referring to resources like the [Python documentation](https://docs.python.org/3/tutorial/) or other beginner tutorials. Understanding these basics will make working with this project – and Python in general – a lot smoother.

**Comment by cstaelen** [comment]
Hi @oskvr37  👋

Few questions if you please :
Is there a way to download videos ? 
How complicated would it be to implement this kind of feature ? 
Is there any chance to see this functionality in 2.0 ?

Btw, great work 👏 
Works like a charm 👌

**Reply by oskvr37** [reply]
Hi!
I haven’t researched downloading videos in Tidal, but I am sure there is easy way to download them.
It will probably be added in future releases of the 2.0 version!

**Reply by cstaelen** [reply]
Thanks for your reply and for your work  💪

**Reply by cstaelen** [reply]
Should I create an issue tagged "enhancement" for this ?

**Comment by cisko99za** [comment]
i hope in next rel you can add again download progress bar

**Reply by oskvr37** [reply]
The downloading process will be redesigned, there will be such features as total progress, current downloading track progress bar and other

**Comment by cisko99za** [comment]
when? i have some plist to download, but i waiting for download bar. i need it for some reasons
could you give a time frame? 1 day? 1 week? 1 month? thanks.

**Reply by Darkloke** [reply]
I suppose a lot of features will be added here, when Tidal will finally release its API 2.0 (or at least i hope for this). :)

**Reply by oskvr37** [reply]
There already was the progress bar in some tiddl version.
You can search for it and then implement it in your code locally

**Comment by cisko99za** [comment]
yes, thanks, i know. and already tried some old version for that, but can't find the version of image on the home page refers to `https://raw.githubusercontent.com/oskvr37/tiddl/refs/heads/main/docs/demo.gif`
this would be enough for me (also because if you wait for API 2.0 from tidal who knows when it will be available...)

**Comment by cisko99za** [comment]
found. version 1.4.0 from download.py
def showProgressBar

now i need to merge ProgressBar into latest 2.3.3 version ...!

**Comment by cisko99za** [comment]
nothing work. tried to edit file download.py and add def showProgressBar but always load symbol to left right like your original script do ...

**Comment by Ultrajackstr** [comment]
Hi,

Thank you for your app. Would it be possible to add a "{today-date}" formatting key ? It would be useful when downloading, for example, the "My Daily Discovery" playlist. Currently, the tracks all go in the same "My Daily Discovery" folder, regardless of when the playlist is downloaded. As it changes everyday by design, this would allow the creation of a different folder for each playlist download. It is not an ideal solution as it would also add the date to all the other playlists, but it would be nice to have the option 😊.

**Reply by oskvr37** [reply]
Hi!
You can utilize the `date` command in CLI.

```sh
tiddl url [playlist url] download -o "daily_discovery/$(date +%Y-%m-%d)/{title}"
```

**Reply by Ultrajackstr** [reply]
Hi, yes, this is actually what I was thinking of right after writing my message, thank you ! It may be something to implement in Tidarr for you, @cstaelen 😊

**Reply by cstaelen** [reply]
@Ultrajackstr what would be the best way to handle this option in tidarr ? 
Should it be only for mixes ?
You can open an issue on tidarr project ;)

**Comment by esturniolo** [comment]
Hello @oskvr37!

What about to add a config option like `singles_filter` but for Comoilation and Live albums?
So we can get all the studio albums instead of everything. 

Thanks!

**Reply by oskvr37** [reply]
Hi!

I'm not sure if we can distinguish compilations and live albums programmatically with data that Tidal provides. I think the only way is to filter these by hand or to set a regex filter.

I will be working on adding `export` command to tiddl - user will be able to process data by himself (filter out resources etc.) and then download the data with tiddl.

**Reply by Darkloke** [reply]
> I'm not sure if we can distinguish compilations and live albums programmatically with data that Tidal provides. I think the only way is to filter these by hand or to set a regex filter.

I dont know if its possible in Tidal, but as i know such thing is possible in yandex music. [This](https://github.com/llistochek/yandex-music-downloader) tool supports two arguments:
 --stick-to-artist     Load albums, only made by artist
 --only-music          Load only music (skip podcasts and audiobooks for example)
 May be something similiar exist in Tidal API too?

**Reply by oskvr37** [reply]
> > I'm not sure if we can distinguish compilations and live albums programmatically with data that Tidal provides. I think the only way is to filter these by hand or to set a regex filter.
> 
> I dont know if its possible in Tidal, but as i know such thing is possible in yandex music. [This](https://github.com/llistochek/yandex-music-downloader) tool supports two arguments: --stick-to-artist Load albums, only made by artist --only-music Load only music (skip podcasts and audiobooks for example) May be something similiar exist in Tidal API too?

These are two completely different things.

**Reply by esturniolo** [reply]
Oh... bad boy Tidal, bad boy! 
XD

At least we tried.

**Reply by oskvr37** [reply]
If you can provide me some examples of live albums/compilations from Tidal, I can try to look up if there are any fields from Tidal API that tell the album type.

Currently there are only 3 types documented

https://github.com/oskvr37/tiddl/blob/a8e3876dfda960f3316c31f18ab125635e3493cb/tiddl/core/api/models/resources.py#L127

**Reply by esturniolo** [reply]
Sadly I think that there is no way to differentiate between studio albums and live ones.

I tried their API with "Live" of ACDC (id: 35986191) and the response is  (I added some notes in specific lines):

```
{
  "data": {
    "attributes": {
      "accessType": "PUBLIC",
      "availability": [
        "STREAM",
        "DJ"
      ],
      "barcodeId": "886444890120",
      "copyright": {
        "text": "(P) 1992 Leidseplein Presse B.V."
      },
      "duration": "PT1H11M3S",
      "explicit": false,
      "externalLinks": [
        {
          "href": "https://tidal.com/browse/album/35986191",
          "meta": {
            "type": "TIDAL_SHARING"
          }
        }
      ],
      "mediaTags": [
        "HIRES_LOSSLESS",
        "LOSSLESS"
      ],
      "numberOfItems": 14,
      "numberOfVolumes": 1,
      "popularity": 0.4785759110934661,
      "releaseDate": "1992-10-27",
      "title": "Live",
      "type": "ALBUM"                                         <--------------------------------------------- ALBUM
    },
    "id": "35986191",
    "relationships": {
      "artists": {
(...)
    },
    "type": "albums"                            <--------------------------------------- ALBUM
  },
  "included": [
(..)
}
```

With this I see that Tidal handle live and studio albums as `ALBUMS` so in short... we are f*cked. 
(I think)

**Reply by oskvr37** [reply]
In general - Tidal has many weird practices if it comes to their backend.

Im thinking of integrating tiddl with some python package that would provide more detailed metadata.

**Comment by Darkloke** [comment]
I have a question about discography download... I launched such download and so far the total progress timer looks like this:
<img width="1137" height="109" alt="image" src="https://github.com/user-attachments/assets/4abe930b-ec37-4f0e-84a7-598b9bd1ea70" />
I suppose the tool does some "caching" before actual download, but what will happen in case of network disconnect or such? Will it start over again? May be in this case there is a reason to make it segmented? Like let cache 5/10/15 albums and then continue?

And it seems in my case it failed to finish:
<img width="1162" height="149" alt="image" src="https://github.com/user-attachments/assets/c1dafd9d-108e-4ecf-be3e-9c9de2a6b5ad" />

**Reply by oskvr37** [reply]
Just run previous command again. Most of API Endpoints are cached for an hour.

**Reply by Darkloke** [reply]
`Error: Response payload is not completed: <ContentLengthError: 400, message='Not enough data to satisfy content length header.'>`
This error is also related to network? I suspect in my case i spend a lot time on caching, but de facto getting not much albums due to error. And then i have to restart the cycle again. This could make downloading a large discography too problematic to me...
<img width="1892" height="121" alt="image" src="https://github.com/user-attachments/assets/b4c7cb4c-67c8-4956-9326-2a0e2cb448e1" />
When the current timer is already more then the hour, could it be that the earlier cachings are already outdated and restarting the cycle starts from the beginning again?

I suppose my vpn could be sometimes unreliable in terms of dns resolving:
<img width="1846" height="117" alt="image" src="https://github.com/user-attachments/assets/7c5750a2-ee68-4cfc-8469-16ec0c7a5d47" />

**Reply by oskvr37** [reply]
Yes, this error is related to the network.
If timer is running for so long and there is 0/0 in total progress, then there are no items added to queue -> network is failing I guess.
Check the logs and see what exactly is happening - whether resources are fetched or not etc.

**Comment by Darkloke** [comment]
Ok, two more questions...

1. Could we run a batch/queue downloads, for example, by linking a text file, which contains download links? Sorry if its already exist and i just missed it.
2.  What settings (compression) used for converting to FLAC? Can we operate them now?

**Reply by oskvr37** [reply]
1. Such function was available in 2.0, I will have to implement it again.
2. There is no converting to FLAC, we are extracting the flac stream from m4a container, Tidal stores max quality tracks flac stream in m4a containers.

**Comment by Darkloke** [comment]
Another idea came to mind...

Tiddl has a function to skip existing downloaded content and/or update only changed metadata. But i assume this only works when files maintain their original path and this matches the one in config file. But lets assume the following situation...

For example i download/listen a lot of production music. On their original sites such music organized in so called "libraries", which have simple and clear structure. Like [this ](https://www.apmmusic.com/libraries/cezame-trailers-(cet))for example. But digital stores dont have such structure and all albums are grouped under various/single artists. So download a few of such libraries and you'll have a mess of albums, which would require some time to manually sort them. But after such sorting you'll lose their original path and thus wont be able to "update" them.

So an idea - may be its possible to save along with the album some information about its source link (like album id, etc.), so you can freely move/sort albums after download and still have an ability to update them. So it would be some kind of recursive folder checking when tiddl scan directories to find such info files to form a list of "download updates". Its somehow similar to torrent recheck/rehash feature, when you scan what you have and only download missing parts. I suppose in case of database, which stores downloaded data, it could be saved there to excluded unnecessary disk scans.

What do you think?

**Reply by oskvr37** [reply]
> So an idea - may be its possible to save along with the album some information about its source link (like album id, etc.),

There is already an idea of item tracking database #211

> so you can freely move/sort albums after download and still have an ability to update them

Tiddl would still need to do a full scan of directory to update database with new locations of items.

My idea is to store items in following format: `base_directory/{item.id}`.
We would have every item in single directory -> fast scans with just one function.
Then user can create symbolic links (symlink) that will *always* link to item id in the base directory.
So when user moves their symlink, then tiddl does not care about it, tiddl always knows where the item source is located - it makes easy updating metadata and tracking already downloaded items.

Then we would have the directory of symlinks: `download_directory/*user_template*`
If user moved a file (symlink) from the download directory, then the source of this symlink wouldn't be changed and tiddl would be fine with it.

And I think that it is future destination for tiddl. Usage of symbolic links would make huge difference in using tiddl - for developer and users.

*edit*

An issue with symlinks is that we can't see metadata of the symlink on Windows (idk how it is on Linux).
We can use hardlinks - then metadata is visible but we can't hardlink source from partition A to partition B. So user would need to store their base directory with download directory at the same partition.

**Reply by Darkloke** [reply]
I see, should be a working variant too. I only hope that it wouldnt create some mess/error when you try to move/sort items.

**Comment by EldonMcGuinness** [comment]
I love the idea of being able to export the actual Tidal data and then parse it myself, will this export also have the ability to export a list of artists I'm following?

**Comment by run51** [comment]
hi


Hello, as with Tidal DL NG, would it be possible to add an option to download files in Atmos format?

Thank you for your great work
